Add CPU Pinning document (#639)
harvester/docs: docs/vm/cpu-pinning.md

Example (node with 16 CPU cores):
```
systemReserved: 408 millicores
kubeReserved: 612 millicores
```
In this example, approximately 15 cores (14980 millicores) are available for workloads.

When a VM (pod) in the Guaranteed QoS class requests 4 CPUs, 4 cores are dedicated to that VM. Pods in the other QoS classes share the remaining 11 cores in the shared pool.

## Enable and Disable CPU Manager

When you enable the CPU Manager, Harvester sets the CPU Manager policy to `static`. When you disable the feature, Harvester switches the CPU Manager policy back to `none`. 

You must enable or disable the CPU Manager on each node separately.

1. On the Harvester UI, go to **Hosts**.

1. Locate the node in the list, and then select **⋮** > **Enable CPU Manager** or **Disable CPU Manager**.

Allow some time for Harvester to apply the corresponding CPU Manager policy.

![enable-cpu-manager](/img/v1.4/cpu-pinning/enable-cpu-manager.png)
![disable-cpu-manager](/img/v1.4/cpu-pinning/disable-cpu-manager.png)

### Limitations

- The CPU Manager cannot be enabled on the witness node.

> - You can enable or disable the CPU Manager on only one management node at a time.

Reviewer: This sounds like when a user enables CPU manager on a management node, he/she can't enable another management node. It would be nice to re-phase it more clearly.

> - You can enable or disable the CPU Manager on only one management node at a time.
> 
> - You must wait for an operation (enabling or disabling) to be completed before starting another.

Reviewer 2: ```suggestion
- The CPU Manager must be enabled or disabled on each node separately. You must wait for the operation to be completed before starting another.
```

Reviewer 2: @brandboat Let's combine points 2 and 3 to address @bk201 's comment.

- VMs with CPU pinning enabled must be stopped before CPU Manager is disabled on the corresponding node.

## Enable CPU Pinning on a New VM

1. Verify that CPU Manager is enabled on one or more nodes.

1. Go to **Virtual Machines**.

1. Click **Create**.

  ![create-vm](/img/v1.4/cpu-pinning/create-vm.png)

1. On the **Advanced Options** tab, select **Enable CPU Pinning**.

  ![vm-advanced-options](/img/v1.4/cpu-pinning/vm-advanced-options.png)

> 1. On the **Advanced Options** tab, select **Enable CPU Pinning**.

Reviewer 2: The step was somehow duplicated. Please remove these lines.

  ![vm-advanced-options](/img/v1.4/cpu-pinning/vm-advanced-options.png)

1. Click **Save**.

Enabling CPU pinning adds `dedicatedCpuPlacement: true` to `.spec.template.spec.domain.cpu` in the virtual machine configuration (YAML). When `dedicatedCpuPlacement` is set to `true`, the CPU and memory resource requests are automatically set to match the limits to ensure that the criteria for Guaranteed QoS are met.

Because the requests and limits are identical, CPU and memory [Resource Overcommit](./resource-overcommit.md) settings do not apply to VMs with CPU pinning enabled.

:::note

To use CPU pinning on an existing VM, you must restart the VM after enabling the feature and saving the change.

:::

## VM Live Migration

VMs with CPU pinning enabled can be migrated only if the CPU Manager is enabled on the target node. For more information, see [Live Migration](./live-migration.md).

## Upgrades

> When upgrading a node, Harvester drains all pods and live migrates VMs to another node. To avoid interruptions to the upgrade process, ensure that the CPU Manager is enabled on other nodes and sufficient resources available whenever you use VMs with CPU pinning enabled.

Reviewer 2: ```suggestion
When upgrading a node, Harvester drains all pods and live migrates VMs to another node. To avoid interruptions to the upgrade process, ensure that the CPU Manager is enabled on other nodes and sufficient resources are available whenever you use VMs with CPU pinning enabled.
```

For instance, on a node with 16 CPU cores, this translates to:
```
systemReserved: 408 millicores
kubeReserved: 612 millicores
```
This leaves ~15 cores (14980 millicores) available for workloads.

When a VM (pod) requests 4 CPUs with Guaranteed QoS, those 4 cores are dedicated to it. The remaining pods that are not in Guaranteed QoS class share the remaining CPU pool (~11 cores).

## How to Enable/Disable CPU Manager ?

Enabling the CPU Manager sets the CPU manager policy to static. Disabling the CPU Manager switchs the policy back to the none policy. This action must be performed individually on each node.

1. Navigate to the Hosts page.
2. Click the hamburger button.
3. Select either Enable CPU Manager or Disable CPU Manager.
4. Wait ~1 minute for the action to complete.

![enable-cpu-manager](/img/v1.4/cpu-pinning/enable-cpu-manager.png)
![disable-cpu-manager](/img/v1.4/cpu-pinning/disable-cpu-manager.png)

### Limitations

> - Only one master node can enable/disable the CPU Manager at a time.

Reviewer 3: we need to rephrase this, since the CPU manager can be enable on any node in the cluster except the witness node

Author: Thanks, already refine the limitations part, also add description about what harvester behaves after user checks the `enable CPU pinning` checkbox in virtual machine page. As you mentioned offline, this would help user understand how to enable cpu pinning via YAML.

For instance, on a node with 16 CPU cores, this translates to:
```
systemReserved: 408 millicores
kubeReserved: 612 millicores
```
This leaves ~15 cores (14980 millicores) available for workloads.

When a VM (pod) requests 4 CPUs with Guaranteed QoS, those 4 cores are dedicated to it. The remaining pods that are not in Guaranteed QoS class share the remaining CPU pool (~11 cores).

## How to Enable/Disable CPU Manager ?

Enabling the CPU Manager sets the CPU manager policy to static. Disabling the CPU Manager switchs the policy back to the none policy. This action must be performed individually on each node.

1. Navigate to the Hosts page.
2. Click the hamburger button.
3. Select either Enable CPU Manager or Disable CPU Manager.
4. Wait ~1 minute for the action to complete.

![enable-cpu-manager](/img/v1.4/cpu-pinning/enable-cpu-manager.png)
![disable-cpu-manager](/img/v1.4/cpu-pinning/disable-cpu-manager.png)

### Limitations

- Only one master node can enable or disable the CPU Manager at a time. Note that the witness node cannot perform these actions.
- You must wait for an operation (enable/disable) to finish before starting another on the same node.

> - Any VMs with CPU pinning enabled must be stopped before disabling CPU Manager on that node.

Reviewer 3: CPU manager is not limited to management/controlplane node, it can be enable on any node in the cluster. We shoudl rephrase this to better reflect this

`You can enable or disable the CPU Manager on only one management node at a time.`

> - Any VMs with CPU pinning enabled must be stopped before disabling CPU Manager on that node.

Author: Thanks for the comment @ibrokethecloud ! What I want to emphasize here is that users could only enable/disable cpu manager  on one management node **_at the same time_.** 

Just a note, I add this limitation since enable/disable cpu manager requires restarting rke2-server on management node, if all rke2-server are restarting in the same time, that would cause whole cluster unavailable.

## How to create a VM with CPU Pinning ?

1. Ensure at least one node enable CPU Manager.
2. Go to Virtual Machines page, click Create. 
3. Click Advanced Options.
![create-vm](/img/v1.4/cpu-pinning/create-vm.png)
4. Scroll down to the bottom, check the Enable CPU Pinning checkbox.
![vm-advanced-options](/img/v1.4/cpu-pinning/vm-advanced-options.png)

> Enabling the checkbox adds `dedicatedCpuPlacement: true` to `.spec.template.spec.domain.cpu` in the virtual machine YAML. In Harvester, when we detect that `dedicatedCpuPlacement` is set to true in the virtual machine definition, we automatically set the CPU and memory resource requests equal to the limits to ensure it meets the criteria for Guaranteed QoS.
> 
> - To enable CPU pinning for an existing VM, restart the VM after enabling this option to apply the change.
> - CPU and memory [Resource Overcommit](./resource-overcommit.md) does not apply to VMs with CPU pinning, as they use Guaranteed QoS, meaning their requests and limits are the same.

Reviewer 2: ```suggestion
Enabling CPU pinning adds `dedicatedCpuPlacement: true` to `.spec.template.spec.domain.cpu` in the virtual machine configuration (YAML). When `dedicatedCpuPlacement` is set to `true`, the CPU and memory resource requests are automatically set to match the limits to ensure that the criteria for Guaranteed QoS are met.

Because the requests and limits are identical, CPU and memory [Resource Overcommit](./resource-overcommit.md) settings do not apply to VMs with CPU pinning enabled.

:::note

To use CPU pinning on an existing VM, you must restart the VM after enabling the feature and saving the change.

:::
```

> ## VM Live Migration
> The migration of a VM with CPU pinning is mentioned in [Live Migration](./live-migration.md). However, you must ensure that the target node has CPU manager enabled; otherwise, the migration will not succeed.
> 
> ## Upgrade
> During an upgrade, Harvester will drain all pods on the node and live migrate VMs to another node. If there are VMs with CPU pinning enabled, ensure that the other nodes have the CPU Manager enabled. Otherwise, the upgrade process may get stuck.

Reviewer 2: ```suggestion
## VM Live Migration

VMs with CPU pinning enabled can be migrated only if the CPU Manager is enabled on the target node. For more information, see [Live Migration](./live-migration.md).

## Upgrades

When upgrading a node, Harvester drains all pods and live migrates VMs to another node. To avoid interruptions to the upgrade process, ensure that the CPU Manager is enabled on other nodes whenever you use VMs with CPU pinning enabled.
```

Author: I changed the sentence a bit:

When upgrading a node, Harvester drains all pods and live migrates VMs to another node. To avoid interruptions to the upgrade process, ensure that the CPU Manager is enabled on other nodes and sufficient resources available whenever you use VMs with CPU pinning enabled.

Only added `and sufficient resources available` here.

## VM Live Migration

> VMs with CPU pinning enabled can be migrated only if the CPU Manager is enabled on the target node.

Author: I removed the reference to `For more information, see [Live Migration](./live-migration.md).` to avoid misleading users into thinking we provide additional information about live migration with CPU manager enabled, which we currently do not.
